added tests for lesson 4 exercises and updated test messages

AnnaDinaburgVulikh opened the pull request:
Hi @shiramax, I have added tests for lesson 4 - lists and dictionaries.
I also amended the tests so they show the following:
1. message the user when the code/output/both been tested.
2. prints all the errors of the exercise. (It only printed the first one before)

shiramax commented:
Hi @AnnaDinaburgVulikh thanks for your contribution!
I'm running  python rose_check.py -e
and I can't see dictionaries and lists in the available tests, am I missing something?

AnnaDinaburgVulikh commented:
Hi @shiramax, the tests work without the latest update.
I missed the update of the pytest.ini file  for the readme, fixed.

shiramax commented on docs/course_materials/exercises/test_exercises/04_Compound_data_types/test_homework_dictionary.py:
```diff
@@ -0,0 +1,72 @@
+        # Add your family members names to the dictionary with their ages.
+        # for example: “Samira”:8
+        [r'''(ages\.update\({)|(ages\[.*\]=\d*)''',
+         'add a member using dict.update() or dict(key)=value'],
```
also a good student work that failes:
ages = {}
ages = {'Ran' : 50, 'Adi' : 49, 'Oren' : 22, 'Shira' : 20, 'Erez' : 17}
print (ages)
print((ages['Oren'] + ages['Shira'] + ages['Erez'] + ages['Ran'] + ages['Adi'])/5)

failures: 
-------------------------------------------------- live log call --------------------------------------------------
[    INFO] Started testing /home/mshira/Downloads/Erez/family_members_average_age.py:
[    INFO] testing the code...
[ WARNING] 
add a member using dict.update() or dict(key)=value
[    INFO] testing the output...
FAILED                                                                                                      [100%]

also- is it possible to check here if the student use len(ages) for calculating the average?

AnnaDinaburgVulikh replied:
Setting the dictionary - this solution works, but wasn't the point to use update or assigning, maybe we should add a hint in the exercise? Also, if I understand correctly, the command used just rewrites the variable.
Len - the regex is set to allow both: len() and a number.
Please let me know which option you prefer for those cases.

AnnaDinaburgVulikh replied:
fixed

shiramax commented on docs/course_materials/exercises/test_exercises/04_Compound_data_types/test_homework_lists.py:
```diff
@@ -0,0 +1,108 @@
+        [r'''^[#]+.*((C|c)reate|1)''',
+         'comment for exercise number 1 is missing'],
+        # Create two empty lists list1 and list2 in different ways
+        [r'''^list.*\s*=\s*\[\]''', 'use [] to set a list'],
```
The error messages here are helpful for the student, but I'm not sure how much helpful they should be, I rather tell the student what they did wrong and not how they need to fix it

AnnaDinaburgVulikh replied:
Do you mean all the answers in this exercise or only the two above?

shiramax replied:
all the exercise

AnnaDinaburgVulikh replied:
Done

shiramax reviewed (changes requested):
names.py: 
my_name = "Yevgeny"
my_last_name = "Gorbachev"
my_full_name = my_name + " " +  my_last_name
my_city_name = "Kyriat Malachi"
print("My name is %s  and I\'m from %s" % (my_full_name, my_city_name))

the error I'm getting which are not completely true: 
[ WARNING] 
my_family_name variable is missing
my_full_name variable is not defined
my_city variable is missing
my_message variable is not defined

AnnaDinaburgVulikh commented:
> names.py:
> my_name = "Yevgeny"
> my_last_name = "Gorbachev"
> my_full_name = my_name + " " + my_last_name
> my_city_name = "Kyriat Malachi"
> print("My name is %s and I'm from %s" % (my_full_name, my_city_name))
> 
> the error I'm getting which are not completely true:
> [ WARNING]
> my_family_name variable is missing
> my_full_name variable is not defined
> my_city variable is missing
> my_message variable is not defined

In this case, the code gets the result, but it's not following the defined instructions in the exercise.
The exercise states "Create a string variable with your name, call it my_name", so if there is no 'my_name' variable it gives an error.
Also, according to the changes made to the course materials only f'string' should be used for formatting while this example uses other formatting expression. 
I can adapt to any of those, just need the boundaries defined.
(Noticed that the last variable is not in the convention, will be fixed to my_message)

AnnaDinaburgVulikh commented on docs/course_materials/exercises/test_exercises/04_Compound_data_types/test_homework_dictionary.py:
```diff
@@ -0,0 +1,72 @@
+        # "Italy":2 "Spain":3 "Israel":1
+        [r'''^.*\s*=\s*\{.*\bItaly\b.*\:\s?2\s?,.*\bSpain\b.*\:\s?3\s?,''' +
+         r'''.*\bIsrael\b.*\:\s?1\s?\}''',
+         'set the dictionary using: {"Italy":2 "Spain":3 "Israel":1}'],
```
Did the needed adjustments.

AnnaDinaburgVulikh replied:
fixed

AnnaDinaburgVulikh commented:
Hi @shiramax, Fixed all requested changes and added a check for 'not best' solution. Used it for the f' check and the len()

shiramax commented on docs/course_materials/exercises/test_exercises/04_Compound_data_types/test_homework_lists.py:
```diff
@@ -0,0 +1,109 @@
+
+    helpers.expected_stdout = [
+        [r'\b4\b[^\]]$',
+         '1.3. Index of number 6 is incorrect'],
```
a student work: 
l1=[2,3,4,5,6]
print(l1.index(6))
ans she still got: 
1.3. Index of number 6 is incorrect

shiramax commented on docs/course_materials/exercises/test_exercises/04_Compound_data_types/test_homework_lists.py:
```diff
@@ -0,0 +1,109 @@
+        [r'\b7\b[^\]]$',
+         '1.5. length of l1 is incorrect'],
+        [r'\[1, 2, 3, 4\]',
+         '2.1. Expected the first 4 numbers'],
```
for x in range(0,4):
	print(l1[x])
should work (not in the way we wanted) but still fails.

shiramax commented on docs/course_materials/exercises/test_exercises/04_Compound_data_types/test_homework_lists.py:
```diff
@@ -0,0 +1,109 @@
+        [r'\[1, 2, 3, 4\]',
+         '2.1. Expected the first 4 numbers'],
+        [r'\[5, 6, 7, 8\]',
+         '2.2. Expected the last 4 numbers'],
```
for x in range(4,8):
	print(l1[x])

same here

and I will also add in the warning message something like "Expected to a print of the last 4 numbers: 5,6,7,8

AnnaDinaburgVulikh commented on docs/course_materials/exercises/test_exercises/conftest.py:
```diff
@@ -103,21 +114,43 @@ def test_answers(expected_list, answer_list, word_pattern=False):
+                    message += '\n' + answer[1]
+        # assert len(message) == 0, LOGGER.warning(message)
+        if len(message) != 0:
+            LOGGER.info(f'Good job, but needs some work:\n{message}')
```
It's different tests, the first paragraph checks only syntax, the second one tests the output.
There are some exercises that both are checked.

AnnaDinaburgVulikh replied:
Fixed

AnnaDinaburgVulikh commented on docs/course_materials/exercises/test_exercises/04_Compound_data_types/test_homework_lists.py:
```diff
@@ -0,0 +1,109 @@
+        [r'\[1, 2, 3, 6, 7, 8, \-1, \-2, \-3\]',
+         '3.3. Expected the l3 list that contains l1 and l2'],
+        [r'\[\-3, \-2, \-1, 1, 2, 3, 6, 7, 8\]',
+         '3.4. Expected a sorted l3 list'],
```
it's 3.4 and it's from the wrong use of sort

shiramax replied:
I can understand that, but is there any way we can send the correct warning here instead of ERROR?

AnnaDinaburgVulikh replied:
Found the problem, will be fixed

AnnaDinaburgVulikh commented on docs/course_materials/exercises/test_exercises/04_Compound_data_types/test_homework_lists.py:
```diff
@@ -0,0 +1,109 @@
+         '2. Expected a comment for exercise 2 - Slicing a list'],
+        # Print first 4 numbers of l1
+        [r'''(print\()?l1\[0?:4\]\)?''',
+         '2.1. Expected the use of range for printing a sequence'],
```
Do you want to add this option as an answer? (the for loop with range). Will fix the message

AnnaDinaburgVulikh commented:
Hi @shiramax , all the changes were made. 
Added the range as a viable option with 'better solution' comment'.
Notice that the exercise is checked twice, once for the syntax and once for the actual output.
Some of the questions pop as errors twice, but the check is different.

shiramax commented on docs/course_materials/exercises/test_exercises/04_Compound_data_types/test_homework_dictionary.py:
```diff
@@ -0,0 +1,81 @@
+         'define a dictionary using `ages = {}`'],
+        # Add your family members names to the dictionary with their ages.
+        # for example: “Samira”:8
+        [r'''(ages\.update\({)|(ages\[.*\]=\d*)''',
```
The description should be "Expected.."

AnnaDinaburgVulikh replied:
Fixed

AnnaDinaburgVulikh commented on docs/course_materials/exercises/test_exercises/04_Compound_data_types/test_homework_dictionary.py:
```diff
@@ -0,0 +1,81 @@
+        # and print it.
+        [r'''(for.*in ages\.values\(\):)|(sum\(ages\.values\(\)\))|''' +
+         r'''(ages\[.*\]\s*\+\s*ages\[.*\])''',
+         'sum the members by using: sum(dict.values() or ' +
```
Fixed

AnnaDinaburgVulikh commented on docs/course_materials/exercises/test_exercises/pytest.ini:
```diff
@@ -3,6 +3,8 @@ addopts = --tb=no -v --no-header --no-summary
     variables: select tests for lesson 03 - variables homework
     strings: select tests for lesson 03 - strings homework
+    lists: select tests for lesson 04 - lists homework
+    dictionary: select tests for lesson 04 - dictionary homework
```
Fixed

AnnaDinaburgVulikh commented:
Hi @shiramax , ammended the logic of the tests to check the output and code together. Please see if you get the wanted result - **right now can be applied only to the lists exercise**.
If the new logic is compatible with the requirement I will ammend the README files and the other exercises to follow.
Also, not sure about the following:
1. Adding the exercise name for each error? Not sure what you meant.
2. Giving feedback for definig a list as strings instead of integers - ammended the error text, not sure if another check can be made.

shiramax commented on docs/course_materials/exercises/test_exercises/04_Compound_data_types/test_homework_lists.py:
```diff
@@ -0,0 +1,115 @@
+         '3. Expected a comment for exercise 3 - Update the list'],
+        # Remove the items 4 and 5 from the l1 and print l1
+        [r'''(del l1\[3:5\])|(l1.remove\(4\))''',
+         '3.1 Expected the use of `del list[:]` or list.remove(i)'],
```
nitpick: you are missing a dot here

AnnaDinaburgVulikh replied:
Fixed

shiramax reviewed (changes requested):
Great Job Anna!
1. According to the exercise (https://github.com/RedHat-Israel/ROSE/blob/master/docs/course_materials/exercises/04_Compound_data_types/homework_list.md) There is no a specific exercises 1.2 or 1.3, well they are there but with headlines and  not numbers (1.1 is basically Create a list.1 )so we have two options to make this a bit more clear :
* Edit the exercise file and add those numbers 
* change the output of the errors to contain the headlines instead of numbers. 
either of those is fine by me 
2. ok great

AnnaDinaburgVulikh commented:
Hi @sleviim @sfishbai ,
I've made the update to the exercise that @shiramax requested.
Also refactored the dictionaries test to the new logic, and updated the README accordingly.
Please let me know if this pr needs more attention.

After it's merged I'll refactor the previos exercises to comply with the test logic change.

AnnaDinaburgVulikh commented:
Hi @sleviim , rebased the branch. :)

AnnaDinaburgVulikh commented:
Hi @sleviim ,
fixed all the tests to match the new convention and rebased the branch to current master.
Travis faild on two links:
1. http://ap.hamakor.org.il/2017/ - didn't load on my pc
2. http://ap.hamakor.org.il/2017/tracks.html#main-f1015 - loads but very slowly
Are those links relevant or maybe we can remove them?

sleviim commented:
This PR seems to be huge...
Is there a way you can split it into some smaller parts?

AnnaDinaburgVulikh commented:
Hi @sleviim ,
I have splitted this pr to 3 smaller ones: #429 , #430 , #431
